### PR TITLE
[serve] Add basic calculate_desired_num_replicas function for autoscaling

### PR DESCRIPTION
**Pull request by architkulkarni**
- Adds a stateless function to calculate the desired number of replicas given a list of the (possibly time-averaged) number of ongoing requests per replica.  This input is provided by https://github.com/ray-project/ray/pull/18458/ and we will integrate these in a future PR.
- Adds an optional smoothing factor ("gain") to dampen the scaling
- Adds unit tests

This PR does not yet add the delay feature, which would only scale up or down if the same decision had been made for the last N seconds.  As a result, the autoscaling might be overly sensitive to fluctuations.  However, this is already partially addressed by the time-window averaging in the metrics layer.

In a future PR, we can add this as a method to a stateful class such as the one laid out earlier by @edoakes in `autoscaling_policy.py`.  We can keep track of the result of previous calls to `calculate_desire_num_replicas` to add the delay feature.
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/autoscaling_policy.py [hunk 1]
```diff
@@ -1,6 +1,57 @@
...
+    error_ratio: float = (
+        num_ongoing_requests_per_replica /
+        autoscaling_config.target_num_ongoing_requests_per_replica)
+
+    # Multiply the distance to 1 by the smoothing ("gain") factor (default=1).
+    smoothed_error_ratio = 1 + (
+        (error_ratio - 1) * autoscaling_config.smoothing_factor)
+    desired_num_replicas = math.ceil(
+        current_num_replicas * smoothed_error_ratio)
+
```
Comment (architkulkarni) [inline python/ray/serve/autoscaling_policy.py:47]: Is it appropriate to link to the public autoscaling doc in the comments for an explanation of the algorithm?

File: python/ray/serve/config.py [hunk 2]
```diff
@@ -9,6 +9,37 @@
...
+    downscale_delay_s: float = 600.0
+    # How long to wait before scaling up replicas
+    upscale_delay_s: float = 30.0
+    # Multiplicative "gain" factor to limit scaling decisions
+    smoothing_factor: float = 1.0
+    # The number of replicas to start with when creating the deployment
+    initial_replicas: int = 1
+    # The num_ongoing_requests_per_replica error ratio (desired / current)
+    # threshold for overriding `upscale_delay_s`
+    panic_mode_threshold: float = 2.0
```
Comment (architkulkarni) [inline python/ray/serve/config.py:37]: Out of these options, only `smoothing_factor` is implemented in this PR.

**Review by simon-mo — APPROVED** [review]

**Comment by simon-mo** [comment]
test failing because AutoscalingConfig not part of BackendConfig protobuf. This will be resolved once we merge #18683

**Comment by simon-mo** [comment]
Serve and lint passed.

**Pull request merged**